containers/container_implementation.md

```diff
@@ -29,31 +28,138 @@ Model containers must implement functionality consistent with this information.
+       if socket in receivable_sockets and receivable_sockets[socket] == ZMQ.POLLIN:
+           # The socket has a message to receive
+           socket.recv()
+           msg_type_bytes = socket.recv()
+           msg_type = struct.unpack("<I", msg_type_bytes)[0]
+           assert msg_type == MESSAGE_TYPE_HEARTBEAT
+           heartbeat_type_bytes = socket.recv()
+           heartbeat_type = struct.unpack("<I", heartbeat_type_bytes)[0]
+           assert heartbeat_type == 1
+           break
+   ```
+   
+5. After receiving a heartbeat response from Clipper requesting metadata, the container should send back a **new container message**. This completes the initialization of a session.
 
+### Maintaining a session
+In order to maintain a session, the container should frequently request heartbeats from Clipper as well as poll its sockets for new *heartbeat* and *container content* messages using the following pattern:
+
+1. Maintain a `last_activity` timestamp that will be used to infer whether or not Clipper is still alive. 
+
+2. Poll the container socket for new messages for a predefined duration `D`. If a message is received before `D` elapses, process the message according to its type (*heartbeat* or *container content*) and set `last_activity` to the current time.
+
+3. If `D` elapses, check the time elapsed since `last_activity`. If this exceeds the **socket activity timeout** `TO`, end the session. Otherwise, send a heartbeat message to Clipper.
+
+4. Repeat steps 1-3. 
+
+Python example:
+
+   ```
+   from datetime import datetime
+   last_activity_time = datetime.now()
+   while(True):
+       receivable_sockets = dict(poller.poll(5000)) # Poll the socket for new content for 5 seconds
+       if socket in receivable_sockets and receivable_sockets[socket] == ZMQ.POLLIN:
+           # The socket has a message to receive
+           socket.recv()
+           msg_type_bytes = socket.recv()
+           msg_type = struct.unpack("<I", msg_type_bytes)[0]
+           # Process message according to type
+           ...
+           last_activity_time = datetime.now()
+       else:
+           time_since_activity = (datetime.now() - last_activity_time).seconds
+           if time_since_activity >= 60:
+               # Terminate the connection
+               ...
+           else:
+               send_heartbeat_message(socket)
+   ```
+
+The current implementation uses the values `D = 5 seconds` and `TO = 60 seconds`. 
+
+### Ending a session
+If the container fails to receive a message from Clipper within the time limit specified by the **socket activity timeout**, the session should be terminated by executing the following steps:
+
+1. Unregister the socket from any ZeroMQ pollers.
+
+2. Close the socket.
+
+Python example:
+
+   ```
+   poller.unregister(socket)
+   socket.close()
+   ```
+The container should then attempt to start a new session.
+
+### For additional container lifecycle references, see:

```

Comment:
Very nice writeup. Welcome to the darkside of RPC implementations 😛. People (i.e. me) always start out by thinking that their RPC use case is simple and they should write their own. And then the implementation inevitably needs to get more complex as they realize they need things like heartbeating.